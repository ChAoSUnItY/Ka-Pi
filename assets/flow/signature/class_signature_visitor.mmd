flowchart
    A0[(Entry Point)]
    A1["< Self as FormalTypeParameterVisitable>::visit_parameter_types"]
    A2["< Self as FormalTypeParameterVisitable>::visit_parameter_type"]
    A3[visit_super_class]
    A4[visit_interfaces]
    A5[visit_interface]
    
    A0 --> A1 --> A2 --> A3 --> A4 --> A5