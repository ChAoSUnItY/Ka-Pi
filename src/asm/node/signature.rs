use std::collections::VecDeque;

use serde::{Deserialize, Serialize};

use crate::asm::signature::{
    accept_class_signature_visitor, accept_field_signature_visitor,
    accept_method_signature_visitor, ClassSignatureVisitor, ClassSignatureWriter,
    FieldSignatureVisitor, FieldSignatureWriter, FormalTypeParameterVisitable,
    FormalTypeParameterVisitor, MethodSignatureVisitor, MethodSignatureWriter, TypeVisitor,
};
use crate::error::{KapiError, KapiResult};

/// Data representation of signatures, including [`Class`](Signature::Class), [`Field`](Signature::Field),
/// and [`Method`](Signature::Method).
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum Signature {
    /// Data representation of class signature.
    Class {
        formal_type_parameters: Vec<FormalTypeParameter>,
        super_class: Type,
        interfaces: Vec<Type>,
    },
    /// Data representation of field signature.
    Field { field_type: Type },
    /// Data representation of method signature.
    Method {
        formal_type_parameters: Vec<FormalTypeParameter>,
        parameter_types: Vec<Type>,
        return_type: Type,
        exception_types: Vec<Type>,
    },
}

impl Signature {
    /// Converts signature string into [`Signature::Class`](Signature::Class).
    pub fn class_signature_from_str<S>(string: S) -> KapiResult<Self>
    where
        S: Into<String>,
    {
        let string = string.into();
        let mut collector = ClassSignatureCollector::default();

        accept_class_signature_visitor(&string, &mut collector)?;

        collector
            .signature
            .ok_or(KapiError::ClassParseError(format!(
                "Unable to parse class signature from `{}`",
                string
            )))
    }

    /// Converts signature string into [`Signature::Field`](Signature::Field).
    pub fn field_signature_from_str<S>(string: S) -> KapiResult<Self>
    where
        S: Into<String>,
    {
        let string = string.into();
        let mut collector = FieldSignatureCollector::default();

        accept_field_signature_visitor(&string, &mut collector)?;

        collector
            .signature
            .ok_or(KapiError::ClassParseError(format!(
                "Unable to parse field signature from `{}`",
                string
            )))
    }

    /// Converts signature string into [`Signature::Method`](Signature::Method).
    pub fn method_signature_from_str<S>(string: S) -> KapiResult<Self>
    where
        S: Into<String>,
    {
        let string = string.into();
        let mut collector = MethodSignatureCollector::default();

        accept_method_signature_visitor(&string, &mut collector)?;

        collector
            .signature
            .ok_or(KapiError::ClassParseError(format!(
                "Unable to parse method signature from `{}`",
                string
            )))
    }
}

impl TryFrom<ClassSignatureWriter> for Signature {
    type Error = KapiError;

    fn try_from(value: ClassSignatureWriter) -> KapiResult<Self> {
        Signature::class_signature_from_str(value.to_string())
    }
}

impl TryFrom<FieldSignatureWriter> for Signature {
    type Error = KapiError;

    fn try_from(value: FieldSignatureWriter) -> KapiResult<Self> {
        Signature::field_signature_from_str(value.to_string())
    }
}

impl TryFrom<MethodSignatureWriter> for Signature {
    type Error = KapiError;

    fn try_from(value: MethodSignatureWriter) -> KapiResult<Self> {
        Signature::method_signature_from_str(value.to_string())
    }
}

#[derive(Default)]
struct ClassSignatureCollector {
    signature: Option<Signature>,
    formal_type_parameters: Vec<FormalTypeParameter>,
    super_class: Type,
    interfaces: Vec<Type>,
}

impl ClassSignatureVisitor for ClassSignatureCollector {
    fn visit_super_class(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.super_class = typ))
    }

    fn visit_interface(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.interfaces.push(typ)))
    }

    fn visit_end(&mut self) {
        self.signature = Some(Signature::Class {
            formal_type_parameters: self.formal_type_parameters.clone(),
            super_class: self.super_class.clone(),
            interfaces: self.interfaces.clone(),
        })
    }
}

impl FormalTypeParameterVisitable for ClassSignatureCollector {
    fn visit_formal_type_parameter(
        &mut self,
        name: &str,
    ) -> Box<dyn FormalTypeParameterVisitor + '_> {
        Box::new(FormalTypeParameterCollector::new(
            name.to_owned(),
            |parameter| self.formal_type_parameters.push(parameter),
        ))
    }
}

#[derive(Default)]
struct FieldSignatureCollector {
    signature: Option<Signature>,
    field_type: Type,
}

impl FieldSignatureVisitor for FieldSignatureCollector {
    fn visit_field_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.field_type = typ))
    }

    fn visit_end(&mut self) {
        self.signature = Some(Signature::Field {
            field_type: self.field_type.clone(),
        })
    }
}

#[derive(Default)]
struct MethodSignatureCollector {
    signature: Option<Signature>,
    formal_type_parameters: Vec<FormalTypeParameter>,
    parameter_types: Vec<Type>,
    return_type: Type,
    exception_types: Vec<Type>,
}

impl MethodSignatureVisitor for MethodSignatureCollector {
    fn visit_parameter_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.parameter_types.push(typ)))
    }

    fn visit_return_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.return_type = typ))
    }

    fn visit_exception_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.exception_types.push(typ)))
    }

    fn visit_end(&mut self) {
        self.signature = Some(Signature::Method {
            formal_type_parameters: self.formal_type_parameters.clone(),
            parameter_types: self.parameter_types.clone(),
            return_type: self.return_type.clone(),
            exception_types: self.exception_types.clone(),
        })
    }
}

impl FormalTypeParameterVisitable for MethodSignatureCollector {
    fn visit_formal_type_parameter(
        &mut self,
        name: &str,
    ) -> Box<dyn FormalTypeParameterVisitor + '_> {
        Box::new(FormalTypeParameterCollector::new(
            name.to_owned(),
            |parameter| self.formal_type_parameters.push(parameter),
        ))
    }
}

/// Data representation of formal type parameter in signatures.
#[derive(Debug, Default, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct FormalTypeParameter {
    parameter_name: String,
    class_bound: Option<Type>,
    interface_bounds: Vec<Type>,
}

struct FormalTypeParameterCollector<F>
where
    F: FnMut(FormalTypeParameter),
{
    parameter_name: String,
    class_bound: Option<Type>,
    interface_bounds: Vec<Type>,
    post_action: F,
}

impl<F> FormalTypeParameterCollector<F>
where
    F: FnMut(FormalTypeParameter),
{
    fn new(parameter_name: String, post_action: F) -> Self {
        Self {
            parameter_name,
            class_bound: None,
            interface_bounds: Vec::new(),
            post_action,
        }
    }
}

impl<F> FormalTypeParameterVisitor for FormalTypeParameterCollector<F>
where
    F: FnMut(FormalTypeParameter),
{
    fn visit_class_bound(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.class_bound = Some(typ)))
    }

    fn visit_interface_bound(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeCollector::new(|typ| self.interface_bounds.push(typ)))
    }

    fn visit_end(&mut self) {
        (self.post_action)(FormalTypeParameter {
            parameter_name: self.parameter_name.clone(),
            class_bound: self.class_bound.clone(),
            interface_bounds: self.interface_bounds.clone(),
        })
    }
}

/// Data representation of Type in signatures.
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum Type {
    BaseType(BaseType),
    Array(Box<Type>),
    Class(String),
    InnerClass(String),
    TypeVariable(String),
    TypeArgument,
    WildcardTypeArgument(Wildcard, Box<Type>),
    /// Unknown is only used in internal code for placeholder usage, you should not see it appears
    /// in returned data structure.
    Unknown,
}

impl Default for Type {
    fn default() -> Self {
        Self::Unknown
    }
}

const EXTENDS: char = '+';
const SUPER: char = '-';
const INSTANCEOF: char = '=';

/// An enum representation for wildcard indicators, which is used in
/// [`Type::WildcardTypeArgument`] as class
/// type argument bound.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[repr(u8)]
pub enum Wildcard {
    /// Indicates type argument must extends class bound, see java's upper bounds wildcard.
    EXTENDS = EXTENDS as u8,
    /// Indicates type argument must super class bound, see java's lower bounds wildcard.
    SUPER = SUPER as u8,
    /// Indicates type argument must be instance of specified type.
    INSTANCEOF = INSTANCEOF as u8,
}

impl Into<char> for Wildcard {
    fn into(self) -> char {
        self as u8 as char
    }
}

impl TryFrom<char> for Wildcard {
    type Error = KapiError;

    fn try_from(value: char) -> KapiResult<Self> {
        match value {
            EXTENDS => Ok(Wildcard::EXTENDS),
            SUPER => Ok(Wildcard::SUPER),
            INSTANCEOF => Ok(Self::INSTANCEOF),
            _ => Err(KapiError::ArgError(format!(
                "Character {} cannot be converted into Wildcard",
                value
            ))),
        }
    }
}

impl TryFrom<&char> for Wildcard {
    type Error = KapiError;

    fn try_from(value: &char) -> KapiResult<Self> {
        TryFrom::<char>::try_from(*value)
    }
}

/// Data representation of base type in descriptor.
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
#[repr(u8)]
pub enum BaseType {
    Boolean = 'Z' as u8,
    Byte = 'B' as u8,
    Short = 'S' as u8,
    Int = 'I' as u8,
    Long = 'J' as u8,
    Float = 'F' as u8,
    Double = 'D' as u8,
    Void = 'V' as u8,
}

impl Into<char> for BaseType {
    fn into(self) -> char {
        self as u8 as char
    }
}

impl TryFrom<char> for BaseType {
    type Error = KapiError;

    fn try_from(value: char) -> KapiResult<Self> {
        match value {
            'Z' => Ok(Self::Boolean),
            'B' => Ok(Self::Byte),
            'S' => Ok(Self::Short),
            'I' => Ok(Self::Int),
            'J' => Ok(Self::Long),
            'F' => Ok(Self::Float),
            'D' => Ok(Self::Double),
            'V' => Ok(Self::Void),
            _ => Err(KapiError::ArgError(format!(
                "Unexpected char `{}` for base type",
                value
            ))),
        }
    }
}

impl TryFrom<&char> for BaseType {
    type Error = KapiError;

    fn try_from(value: &char) -> KapiResult<Self> {
        TryFrom::<char>::try_from(*value)
    }
}

struct TypeCollector<F>
where
    F: FnMut(Type),
{
    holder: Type,
    stack_actions: VecDeque<(Option<Wildcard>)>,
    // If option is None, then it's array wrapping, otherwise it's type argument wrapping
    post_action: F,
}

impl<F> TypeCollector<F>
where
    F: FnMut(Type),
{
    fn new(post_action: F) -> Self {
        Self {
            holder: Type::Unknown,
            stack_actions: VecDeque::new(),
            post_action,
        }
    }

    #[inline]
    fn wrap_array_type(typ: Type) -> Type {
        Type::Array(Box::new(typ))
    }

    #[inline]
    fn wrap_type_argument(wildcard: Wildcard, typ: Type) -> Type {
        Type::WildcardTypeArgument(wildcard, Box::new(typ))
    }
}

impl<F> TypeVisitor for TypeCollector<F>
where
    F: FnMut(Type),
{
    fn visit_base_type(&mut self, base_type: BaseType) {
        self.holder = Type::BaseType(base_type);
    }

    fn visit_array_type(&mut self) {
        self.stack_actions.push_back(None);
    }

    fn visit_class_type(&mut self, name: &str) {
        self.holder = Type::Class(name.to_owned());
    }

    fn visit_inner_class_type(&mut self, name: &str) {
        self.holder = Type::InnerClass(name.to_owned());
    }

    fn visit_type_variable(&mut self, name: &str) {
        self.holder = Type::TypeVariable(name.to_owned());
    }

    fn visit_type_argument(&mut self) {
        self.holder = Type::TypeArgument;
    }

    fn visit_type_argument_wildcard(&mut self, wildcard: Wildcard) {
        self.stack_actions.push_back(Some(wildcard))
    }

    fn visit_end(&mut self) {
        let mut typ = self.holder.clone();

        while let Some(wildcard) = self.stack_actions.pop_back() {
            if let Some(wildcard) = wildcard {
                typ = Self::wrap_type_argument(wildcard.clone(), typ);
            } else {
                typ = Self::wrap_array_type(typ);
            }
        }

        (self.post_action)(typ)
    }
}

#[cfg(test)]
mod test {
    use insta::assert_yaml_snapshot;
    use rstest::rstest;

    use crate::asm::node::signature::Signature;
    use crate::error::KapiResult;

    #[rstest]
    #[case("<T:[Ljava/lang/Object;>Ljava/lang/Object;Ljava/lang/Runnable;")]
    fn test_class_signatures(#[case] signature: &'static str) -> KapiResult<()> {
        let class_signature = Signature::class_signature_from_str(signature)?;

        assert_yaml_snapshot!(class_signature);

        Ok(())
    }

    #[rstest]
    #[case("Ljava/lang/Object;")]
    #[case("TT;")]
    fn test_field_signatures(#[case] signature: &'static str) -> KapiResult<()> {
        let field_signature = Signature::field_signature_from_str(signature)?;

        assert_yaml_snapshot!(field_signature);

        Ok(())
    }

    #[rstest]
    #[case("<T:Ljava/lang/Object;>(Z[[ZTT;)Ljava/lang/Object;^Ljava/lang/Exception;")]
    fn test_method_signatures(#[case] signature: &'static str) -> KapiResult<()> {
        let method_signature = Signature::method_signature_from_str(signature)?;

        assert_yaml_snapshot!(method_signature);

        Ok(())
    }
}
