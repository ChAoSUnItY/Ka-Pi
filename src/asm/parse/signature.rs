use nom::IResult;
use crate::asm::node::signature::Signature;
use crate::asm::node::types::Type;
use crate::error::KapiResult;

pub(crate) fn parse_class_signature() -> KapiResult<Signature> {
    
}

pub(crate) fn java_type_signature(input: &[char]) -> IResult<&[char], Type> {
    
}

pub(crate) fn base_type(input: &[char]) -> IResult<&[char], Ba>
