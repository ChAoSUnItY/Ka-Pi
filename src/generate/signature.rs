use std::collections::VecDeque;

use crate::node::signature::BaseType;
use crate::node::signature::WildcardIndicator;

/// A default implementation of class signature writer.
///
/// This is commonly used in class file generation.
#[derive(Debug, Default)]
pub struct ClassSignatureWriter {
    signature_builder: String,
    has_formal: bool,
}

impl ClassSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
            has_formal: false,
        }
    }

    fn init_formal_types(&mut self) {
        if !self.has_formal {
            self.has_formal = true;
            self.signature_builder.push('<');
        }
    }

    fn finalize_formal_types(&mut self) {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }
    }

    //noinspection DuplicatedCode
    pub fn formal_type_parameter(&mut self, name: &str) -> FormalTypeParameterWriter<'_> {
        self.init_formal_types();

        self.signature_builder.push_str(name);

        FormalTypeParameterWriter::new(&mut self.signature_builder)
    }

    pub fn super_class(&mut self) -> TypeWriter<'_> {
        self.finalize_formal_types();

        TypeWriter::new(&mut self.signature_builder)
    }

    pub fn interface(&mut self) -> TypeWriter<'_> {
        self.finalize_formal_types();

        TypeWriter::new(&mut self.signature_builder)
    }
}

impl ToString for ClassSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

/// A default implementation of field signature writer.
///
/// This is commonly used in class file generation.
#[derive(Debug, Default)]
pub struct FieldSignatureWriter {
    signature_builder: String,
}

impl FieldSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
        }
    }

    pub fn field_type(&mut self) -> TypeWriter<'_> {
        TypeWriter::new(&mut self.signature_builder)
    }
}

impl ToString for FieldSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

#[derive(Debug, Default)]
pub struct MethodSignatureWriter {
    signature_builder: String,
    has_formal: bool,
    has_parameters: bool,
}

/// A default implementation of method signature writer.
///
/// This is commonly used in class file generation.
impl MethodSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
            has_formal: false,
            has_parameters: false,
        }
    }

    fn init_formal_types(&mut self) {
        if !self.has_formal {
            self.has_formal = true;
            self.signature_builder.push('<');
        }
    }

    fn finalize_formal_types(&mut self) {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }
    }

    fn init_parameter_types(&mut self) {
        if !self.has_parameters {
            self.has_parameters = true;
            self.signature_builder.push('(');
        }
    }

    fn finalize_parameter_types(&mut self) {
        if self.has_parameters {
            self.has_parameters = false;
            self.signature_builder.push(')');
        }
    }

    //noinspection DuplicatedCode
    pub fn formal_type_parameter(&mut self, name: &str) -> FormalTypeParameterWriter<'_> {
        self.init_formal_types();

        self.signature_builder.push_str(name);

        FormalTypeParameterWriter::new(&mut self.signature_builder)
    }

    pub fn parameter_type(&mut self) -> TypeWriter<'_> {
        self.finalize_formal_types();
        self.init_parameter_types();

        TypeWriter::new(&mut self.signature_builder)
    }

    pub fn return_type(&mut self) -> TypeWriter<'_> {
        self.finalize_formal_types();
        self.finalize_parameter_types();

        TypeWriter::new(&mut self.signature_builder)
    }

    pub fn exception_type(&mut self) -> TypeWriter<'_> {
        self.signature_builder.push('^');

        TypeWriter::new(&mut self.signature_builder)
    }
}

impl ToString for MethodSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

pub struct FormalTypeParameterWriter<'parent> {
    parent_builder: &'parent mut String,
}

impl<'parent> FormalTypeParameterWriter<'parent> {
    fn new(parent_builder: &'parent mut String) -> Self {
        Self { parent_builder }
    }

    pub fn class_bound(&mut self) -> TypeWriter<'_> {
        TypeWriter::new(&mut self.parent_builder)
    }

    pub fn interface_bound(&mut self) -> TypeWriter<'_> {
        self.parent_builder.push(':');

        TypeWriter::new(&mut self.parent_builder)
    }
}

pub struct TypeWriter<'parent> {
    parent_builder: &'parent mut String,
    type_arg_stack: VecDeque<bool>,
}

impl<'parent> TypeWriter<'parent> {
    fn new(parent_builder: &'parent mut String) -> Self {
        let mut type_arg_stack = VecDeque::with_capacity(64);

        type_arg_stack.push_back(true);

        Self {
            parent_builder,
            type_arg_stack,
        }
    }

    fn end_args(&mut self) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.parent_builder.push('>');
        }

        self.type_arg_stack.pop_front();
    }

    pub fn base_type(&mut self, base_type: BaseType) {
        self.parent_builder.push(base_type.into());
    }

    pub fn array_type(&mut self) {
        self.parent_builder.push('[');
    }

    pub fn class_type(&mut self, name: &str) {
        self.parent_builder.push('L');
        self.parent_builder.push_str(name);
        self.parent_builder.push(';');
        self.type_arg_stack.push_front(false);
    }

    pub fn inner_class_type(&mut self, name: &str) {
        self.end_args();
        self.parent_builder.push('.');
        self.parent_builder.push_str(name);
        self.type_arg_stack.push_front(false);
    }

    pub fn type_variable(&mut self, name: &str) {
        self.parent_builder.push('T');
        self.parent_builder.push_str(name);
        self.parent_builder.push(';');
    }

    pub fn wildcard(&mut self) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.type_arg_stack[0] = true;
            self.parent_builder.push('<');
        }

        self.parent_builder.push('*');
    }

    pub fn type_argument(&mut self, wildcard: WildcardIndicator) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.type_arg_stack[0] = true;
            self.parent_builder.push('<');
        }

        self.parent_builder.push(wildcard.into());
    }

    pub fn finalize(&mut self) {
        self.end_args();
        self.parent_builder.push(';');
    }
}

#[cfg(test)]
mod test {
    use insta::assert_yaml_snapshot;

    use crate::generate::signature::{
        ClassSignatureWriter, FieldSignatureWriter, MethodSignatureWriter,
    };

    #[test]
    fn test_class_signature_writer() {
        let mut writer = ClassSignatureWriter::default();

        writer
            .formal_type_parameter(&"T".to_string())
            .class_bound()
            .class_type("java/lang/Object");

        writer.super_class().class_type("java/lang/Object");
        writer.interface().class_type("java/lang/Comparable");

        assert_yaml_snapshot!(writer.to_string());
    }

    #[test]
    fn test_field_signature_writer() {
        let mut writer = FieldSignatureWriter::default();

        writer.field_type().class_type("java/lang/String");

        assert_yaml_snapshot!(writer.to_string());
    }

    #[test]
    fn test_method_signature_writer() {
        let mut writer = MethodSignatureWriter::default();

        writer
            .formal_type_parameter(&"T".to_string())
            .class_bound()
            .class_type("java/lang/Object");

        writer.parameter_type().class_type("java/lang/Object");
        writer.return_type().class_type("java/lang/String");

        assert_yaml_snapshot!(writer.to_string());
    }
}
