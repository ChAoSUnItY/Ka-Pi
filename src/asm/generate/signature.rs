use std::collections::VecDeque;

use crate::asm::node::signature::{BaseType, Wildcard};
use crate::asm::visitor::signature::{
    ClassSignatureVisitor, FieldSignatureVisitor, FormalTypeParameterVisitable,
    FormalTypeParameterVisitor, MethodSignatureVisitor, TypeVisitor,
};

/// A default implementation of class signature writer.
///
/// This is commonly used in class file generation.
#[derive(Debug, Default)]
pub struct ClassSignatureWriter {
    signature_builder: String,
    has_formal: bool,
}

impl ClassSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
            has_formal: false,
        }
    }
}

impl ToString for ClassSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

impl ClassSignatureVisitor for ClassSignatureWriter {
    fn visit_super_class(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }

    fn visit_interface(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }
}

impl FormalTypeParameterVisitable for ClassSignatureWriter {
    //noinspection DuplicatedCode
    fn visit_formal_type_parameter(
        &mut self,
        name: &str,
    ) -> Box<dyn FormalTypeParameterVisitor + '_> {
        if !self.has_formal {
            self.has_formal = true;
            self.signature_builder.push('<');
        }

        self.signature_builder.push_str(name);

        Box::new(FormalTypeParameterWriter::new(&mut self.signature_builder))
    }
}

/// A default implementation of field signature writer.
///
/// This is commonly used in class file generation.
#[derive(Debug, Default)]
pub struct FieldSignatureWriter {
    signature_builder: String,
}

impl FieldSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
        }
    }
}

impl ToString for FieldSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

impl FieldSignatureVisitor for FieldSignatureWriter {
    fn visit_field_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeWriter::new(&mut self.signature_builder))
    }
}

#[derive(Debug, Default)]
pub struct MethodSignatureWriter {
    signature_builder: String,
    has_formal: bool,
    has_parameters: bool,
}

/// A default implementation of method signature writer.
///
/// This is commonly used in class file generation.
impl MethodSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
            has_formal: false,
            has_parameters: false,
        }
    }
}

impl ToString for MethodSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

impl MethodSignatureVisitor for MethodSignatureWriter {
    fn visit_parameter_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        if !self.has_parameters {
            self.has_parameters = true;
            self.signature_builder.push('(');
        }

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }

    fn visit_return_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        if self.has_parameters {
            self.has_parameters = false;
        } else {
            self.signature_builder.push('(');
        }

        self.signature_builder.push(')');

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }

    fn visit_exception_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        self.signature_builder.push('^');

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }
}

impl FormalTypeParameterVisitable for MethodSignatureWriter {
    //noinspection DuplicatedCode
    fn visit_formal_type_parameter(
        &mut self,
        name: &str,
    ) -> Box<dyn FormalTypeParameterVisitor + '_> {
        if !self.has_formal {
            self.has_formal = true;
            self.signature_builder.push('<');
        }

        self.signature_builder.push_str(name);

        Box::new(FormalTypeParameterWriter::new(&mut self.signature_builder))
    }
}

struct FormalTypeParameterWriter<'parent> {
    parent_builder: &'parent mut String,
}

impl<'parent> FormalTypeParameterWriter<'parent> {
    fn new(parent_builder: &'parent mut String) -> Self {
        Self { parent_builder }
    }
}

impl<'parent> FormalTypeParameterVisitor for FormalTypeParameterWriter<'parent> {
    fn visit_class_bound(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeWriter::new(&mut self.parent_builder))
    }

    fn visit_interface_bound(&mut self) -> Box<dyn TypeVisitor + '_> {
        self.parent_builder.push(':');
        Box::new(TypeWriter::new(&mut self.parent_builder))
    }
}

struct TypeWriter<'parent> {
    parent_builder: &'parent mut String,
    type_arg_stack: VecDeque<bool>,
}

impl<'parent> TypeWriter<'parent> {
    fn new(parent_builder: &'parent mut String) -> Self {
        let mut type_arg_stack = VecDeque::with_capacity(64);

        type_arg_stack.push_back(true);

        Self {
            parent_builder,
            type_arg_stack,
        }
    }

    fn end_args(&mut self) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.parent_builder.push('>');
        }

        self.type_arg_stack.pop_front();
    }
}

impl<'parent> TypeVisitor for TypeWriter<'parent> {
    fn visit_base_type(&mut self, base_type: BaseType) {
        self.parent_builder.push(base_type.into());
    }

    fn visit_array_type(&mut self) {
        self.parent_builder.push('[');
    }

    fn visit_class_type(&mut self, name: &str) {
        self.parent_builder.push('L');
        self.parent_builder.push_str(name);
        self.parent_builder.push(';');
        self.type_arg_stack.push_front(false);
    }

    fn visit_inner_class_type(&mut self, name: &str) {
        self.end_args();
        self.parent_builder.push('.');
        self.parent_builder.push_str(name);
        self.type_arg_stack.push_front(false);
    }

    fn visit_type_variable(&mut self, name: &str) {
        self.parent_builder.push('T');
        self.parent_builder.push_str(name);
        self.parent_builder.push(';');
    }

    fn visit_type_argument(&mut self) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.type_arg_stack[0] = true;
            self.parent_builder.push('<');
        }

        self.parent_builder.push('*');
    }

    fn visit_type_argument_wildcard(&mut self, wildcard: Wildcard) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.type_arg_stack[0] = true;
            self.parent_builder.push('<');
        }

        if wildcard != Wildcard::INSTANCEOF {
            self.parent_builder.push(wildcard.into());
        }
    }

    fn visit_end(&mut self) {
        self.end_args();
        self.parent_builder.push(';');
    }
}

#[cfg(test)]
mod test {
    use insta::assert_yaml_snapshot;

    use crate::asm::generate::signature::{
        ClassSignatureWriter, FieldSignatureWriter, MethodSignatureWriter,
    };
    use crate::asm::visitor::signature::{
        ClassSignatureVisitor, FieldSignatureVisitor, FormalTypeParameterVisitable,
        MethodSignatureVisitor,
    };

    #[test]
    fn test_class_signature_writer() {
        let mut writer = ClassSignatureWriter::default();

        writer
            .visit_formal_type_parameter(&"T".to_string())
            .visit_class_bound()
            .visit_class_type("java/lang/Object");

        writer
            .visit_super_class()
            .visit_class_type("java/lang/Object");
        writer
            .visit_interface()
            .visit_class_type("java/lang/Comparable");

        assert_yaml_snapshot!(writer.to_string());
    }

    #[test]
    fn test_field_signature_writer() {
        let mut writer = FieldSignatureWriter::default();

        writer
            .visit_field_type()
            .visit_class_type("java/lang/String");

        assert_yaml_snapshot!(writer.to_string());
    }

    #[test]
    fn test_method_signature_writer() {
        let mut writer = MethodSignatureWriter::default();

        writer
            .visit_formal_type_parameter(&"T".to_string())
            .visit_class_bound()
            .visit_class_type("java/lang/Object");

        writer
            .visit_parameter_type()
            .visit_class_type("java/lang/Object");
        writer
            .visit_return_type()
            .visit_class_type("java/lang/String");

        assert_yaml_snapshot!(writer.to_string());
    }
}
