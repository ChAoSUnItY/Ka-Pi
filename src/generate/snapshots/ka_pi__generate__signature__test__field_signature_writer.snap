---
source: src/generate/signature.rs
expression: writer.to_string()
---
Ljava/lang/String;

